### PR TITLE
Cleaner attention interfaces

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

Improve the attention interfaces with comments/dtype fix

### CODE REVIEW

File: src/transformers/integrations/sdpa_attention.py [hunk 1]
```diff
@@ -34,10 +48,15 @@ def sdpa_attention_forward(
...
 
-    query = query.contiguous()
-    key = key.contiguous()
-    value = value.contiguous()
+    # SDPA with memory-efficient backend is bugged with non-contiguous inputs and custom attn_mask for some torch versions
+    # Reference: https://github.com/pytorch/pytorch/issues/112577.
+    if _needs_contiguous_inputs:
+        query = query.contiguous()
+        key = key.contiguous()
+        value = value.contiguous()
```
Comment (ArthurZucker) [inline src/transformers/integrations/sdpa_attention.py:56]: we could also just always apply contiguous!

**Review by ArthurZucker — APPROVED** [review]
Cool thanks for updating, thanks @vasqu for the review

**Review by vasqu — COMMENTED** [review]
Just as a heads up, maybe it would be nice to doc what attn type supports what feature, e.g. sdpa and fa cannot output attn weights or use head mask while flex attn cannot use dropout.

Which brings me to the point to ask if these things have been considered (flex attn dropout not supported for example).

**Comment by ArthurZucker** [comment]
LGTM

**Pull request merged**